File: volume1/1_1algorithms/algorithm_e/algorithm_e.cpp
/// The Art of Computer Programming
/// Section 1.1 - Algorithm E
/// Author: Carlos L. Cuenca
/// Date: 10/17/20

#include<iostream>
#include<string>
#include<vector>
#include<cmath>
#include<climits>
#include<algorithm>

/// -----------------
/// Program Constants

const int SUCCESS = 0;

/// -----------------------------
/// Function Forward Declarations

int gcd(int, int);

/// --------------
/// Driver Program

int main(int argc, char* argv[]) {

    /// -----------------
    /// Program Variables

    int m;
    int n;

    std::cin >> m >> n;

    std::cout << gcd(m, n) << std::endl;

	return SUCCESS;

}

/// ----------------------
/// Function Impelentation

int gcd(int m, int n) {
    
    m = std::abs(m);
    n = std::abs(n);

    if(m < n) {

	   	int temp = n;
	   	n = m;
	   	m = temp;

	}

    int remainder = m % n;

    while(remainder) {

	    remainder = m % n;

	    if(!remainder) break;

	    m = n;
	    n = remainder;

    }

    return n;

}

File: volume1/1_1algorithms/algorithm_e/algorithm_e.rb
# Section 1.1 - Algorithm E
# From Knuth's The Art of Computer Programming
# Author: Carlos L. Cuenca
# Date: 10/21/20

# ------------------------
# Function Implementations

def algorithm_e(m, n)

    m = m.abs
    n = n.abs

    if m < n then

        temp = n
        n = m
        m = temp

    end

    remainder = m % n

    while remainder > 0 do

        remainder = m % n

        if remainder == 0 then break end

        m = n
        n = remainder

    end

    return n

end

# -------
# Program

ARGF.each do |line|

    array = line.split.map(&:to_i)

    puts algorithm_e(array[0], array[1])

end
